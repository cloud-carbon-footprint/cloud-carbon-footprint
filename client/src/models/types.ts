/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */
import { Account } from './FilterInputModels'

export interface serviceEstimate {
  cloudProvider: string
  accountName: string
  serviceName: string
  wattHours: number
  co2e: number
  cost: number
  region: string
  usesAverageCPUConstant?: boolean
}

export interface ServiceResult {
  data: EstimationResult[]
  loading: boolean
}

export interface EstimationResult {
  timestamp: Date
  serviceEstimates: serviceEstimate[]
}

export interface cloudEstPerDay {
  x: Date
  y: number
  usesAverageCPUConstant?: boolean
  wattHours?: number
  cost?: number
}

export enum ChartDataTypes {
  REGION = 'region',
  SERVICE = 'service',
  ACCOUNT = 'account',
}

export interface FilterResultResponse {
  accounts: Account[]
}
