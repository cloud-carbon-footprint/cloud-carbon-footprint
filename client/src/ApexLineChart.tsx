import React, { FunctionComponent } from 'react'
import Chart from 'react-apexcharts'
import { transformData } from './transformData'
import { EstimationResult } from './types'

export const ApexLineChart: FunctionComponent<ApexLineChartProps> = ({data}) => {
    const timeSeriesData = transformData(data)
    const options = {
        series: [{
            name: 'co2e',
            data: timeSeriesData
        }],
        chart: {
            id: 'apexchart-example',
        },
        height: '500px',
        legend: {
          show: true
        },
        markers: {
            size: 5,
        },
        stroke: {
            width: 2
        },
        title: {
            text: "Cloud Emissions",
            style: {
                fontSize: '24px'
            },
        },
        xaxis: {
            type: 'datetime',
            title: {
                text: 'Date',
                offsetY: 10,
                style: {
                    fontSize: '15px'
                }
            }
        },
        yaxis:{
            title: {
                text: 'Daily co2e',
                offsetX: -5,
                style: {
                    fontSize: '15px'
                }
            }
        },
    }
    return (
        <Chart options={options} series={options.series} type='line' height={options.height}/>
    )
}

type ApexLineChartProps = {
  data: EstimationResult[]
}
