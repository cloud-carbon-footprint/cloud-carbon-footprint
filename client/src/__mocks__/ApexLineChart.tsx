import React from 'react'

const ApexLineChart = () => <div>fake chart</div>

export { ApexLineChart }
