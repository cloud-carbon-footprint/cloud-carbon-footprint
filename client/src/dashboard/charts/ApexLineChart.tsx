import React, { FunctionComponent } from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import { useTheme } from '@material-ui/core/styles'
import { GetApp, PanTool, RotateLeft, ZoomIn } from '@material-ui/icons'
import Chart from 'react-apexcharts'
import moment from 'moment'

import { getChartColors } from '../../themes'
import { dailyTotals } from '../transformData'
import { EstimationResult } from '../../types'

export const ApexLineChart: FunctionComponent<ApexLineChartProps> = ({ data }) => {
  const theme = useTheme()

  // We need to get the HTML string version of these icons since ApexCharts doesn't take in custom React components.
  // Why, you might ask? Don't ask me, ask ApexCharts.
  const GetAppIconHTML = renderToStaticMarkup(<GetApp />)
  const PanToolIconHTML = renderToStaticMarkup(<PanTool />)
  const RotateLeftIconHTML = renderToStaticMarkup(<RotateLeft />)
  const ZoomInIconHTML = renderToStaticMarkup(<ZoomIn />)

  const co2SeriesData = dailyTotals(data).co2e
  const wattHoursSeriesData = dailyTotals(data).wattHours
  const CostSeriesData = dailyTotals(data).cost

  const colors = getChartColors(theme)
  const [blue, yellow, green] = [colors[0], colors[5], colors[8]]

  const options = {
    chart: {
      background: theme.palette.background.default,
      toolbar: {
        tools: {
          download: GetAppIconHTML,
          zoom: ZoomInIconHTML,
          zoomin: false,
          zoomout: false,
          pan: PanToolIconHTML,
          reset: RotateLeftIconHTML,
        },
      },
    },
    colors: [blue, yellow, green],
    height: '500px',
    markers: {
      size: 5,
    },
    series: [
      {
        name: 'AWS CO2e',
        data: co2SeriesData,
      },
      {
        name: 'AWS Watt Hours',
        data: wattHoursSeriesData,
      },
      {
        name: 'AWS Cost',
        data: CostSeriesData,
      },
    ],
    stroke: {
      width: 2,
    },
    theme: {
      mode: theme.palette.type,
    },
    title: {
      text: 'Cloud Usage',
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    xaxis: {
      type: 'datetime',
      title: {
        text: '',
        offsetY: 8,
        style: {
          fontSize: '15px',
        },
      },
      labels: {
        formatter: function (value: any, timestamp: any, index: any) {
          return moment.utc(timestamp).format('DD-MMM-YY')
        },
      },
    },
    yaxis: [
      {
        title: {
          text: 'CO2e (kg)',
          offsetX: -8,
          style: {
            fontSize: '15px',
          },
        },
        decimalsInFloat: 3,
      },
      {
        title: {
          text: 'Watt hours (Wh)',
          opposite: -8,
          style: {
            fontSize: '15px',
            color: yellow,
          },
        },
        decimalsInFloat: 2,
        opposite: true,
        axisBorder: {
          show: true,
          color: yellow,
        },
        axisTicks: {
          show: true,
        },
      },
      {
        title: {
          text: 'Cost ($)',
          offsetX: -8,
          style: {
            fontSize: '15px',
            color: green,
          },
        },
        decimalsInFloat: 2,
        opposite: true,
        axisBorder: {
          show: true,
          color: green,
        },
        axisTicks: {
          show: true,
        },
      },
    ],
  }
  return <Chart options={options} series={options.series} type="line" height={options.height} />
}

type ApexLineChartProps = {
  data: EstimationResult[]
}
