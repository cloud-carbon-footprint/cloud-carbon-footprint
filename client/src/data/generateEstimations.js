import moment from 'moment'

const getRandomInt = (max) => {
    return Math.floor(Math.random() * Math.floor(max));
}

const getRandomNumberInRange = (minValue, maxValue) => {
    return Math.max(Math.random() * maxValue, minValue + Math.random())
}

const generateEstimations = (today, monthsBack) => {
    const todayAsMoment = moment(today)

    const output = {
        footprint: []
    }

    for (let i = 0; i <= monthsBack; i++) {

        const timestamp = today.clone().utc().subtract(i, 'M').hours(0).minutes(0).seconds(0).millisecond(0)
        const footprintForMonth = {
            timestamp,
            serviceEstimates: [
                {
                    timestamp,
                    serviceName: "ebs",
                    wattHours: Math.random(),
                    co2e: 5,
                    cost: 0
                },
                {
                    timestamp,
                    serviceName: "s3",
                    wattHours: Math.random() / 1000,
                    co2e: 5,
                    cost: 0
                },
                {
                    timestamp,
                    serviceName: "ec2",
                    wattHours: getRandomNumberInRange(50, 75),
                    co2e:  getRandomInt(5),
                    cost: getRandomNumberInRange(1.5, 2)
                }
            ]
        }

        output.footprint.push(footprintForMonth)
    }

    return JSON.stringify(output)
}

export default generateEstimations
  