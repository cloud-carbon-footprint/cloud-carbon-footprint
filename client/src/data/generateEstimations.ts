import moment, { Moment } from 'moment'
import { EstimationResult, serviceEstimate } from '../types'

const getRandomInt = (max: number): number => {
  return Math.floor(Math.random() * Math.floor(max))
}

const getRandomNumberInRange = (minValue: number, maxValue: number): number => {
  return Math.max(Math.random() * maxValue, minValue + Math.random())
}

const fakeServiceMap: (timestamp: Date) => { [key: string]: serviceEstimate } = (timestamp: Date) => {
  return {
    ebs: {
      timestamp,
      serviceName: 'ebs',
      wattHours: Math.random(),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      estimated: false,
    },
    s3: {
      timestamp,
      serviceName: 's3',
      wattHours: Math.random() / 1000,
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      estimated: false,
    },
    ec2: {
      timestamp,
      serviceName: 'ec2',
      wattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      estimated: false,
    },
    rds: {
      timestamp,
      serviceName: 'rds',
      wattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      estimated: false,
    },
    lambda: {
      timestamp,
      serviceName: 'lambda',
      wattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      estimated: false,
    },
    elasticache: {
      timestamp,
      serviceName: 'elasticache',
      wattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      estimated: false,
    },
  }
}

const generateEstimations = (
  today: moment.Moment | Date,
  monthsBack: number,
  servicesToTest: string[] = ['ebs', 's3', 'ec2', 'rds', 'lambda', 'elasticache'],
): EstimationResult[] => {
  const todayAsMoment: moment.Moment = moment(today)

  const estimationResults: EstimationResult[] = []

  for (let i: number = 0; i <= monthsBack; i++) {
    const timestamp: Date = todayAsMoment
      .clone()
      .utc()
      .subtract(i, 'M')
      .hours(0)
      .minutes(0)
      .seconds(0)
      .millisecond(0)
      .toDate()
    const estimationsForMonth: EstimationResult = {
      timestamp,
      serviceEstimates: servicesToTest.map((serviceToTest: string) => {
        return fakeServiceMap(timestamp)[serviceToTest]
      }),
    }

    estimationResults.push(estimationsForMonth)
  }

  return estimationResults
}

export default generateEstimations
