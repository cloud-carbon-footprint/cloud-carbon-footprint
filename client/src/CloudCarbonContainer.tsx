import React, { useState } from 'react'
import useRemoteService from './hooks/RemoteServiceHook'
import { ApexLineChart } from './ApexLineChart'
import moment from 'moment'
import MonthFilter from './MonthFilter'
import CachedIcon from '@material-ui/icons/Cached'
import { Box, Container } from '@material-ui/core'

const CloudCarbonContainer = () => {
  const startDate: moment.Moment = moment.utc().subtract(1, 'year')
  const endDate: moment.Moment = moment.utc()
  const region: string = 'us-east-1'

  const { data, loading } = useRemoteService([], startDate, endDate, region)

  const [dataInTimeframe, setDataInTimeframe] = useState(data)

  return (
    <>
      <Box marginBottom={4}>
        <MonthFilter dataFromRemoteService={data} setDataInTimeframe={setDataInTimeframe} />
      </Box>
      <Box padding={1} border={1}>
        {loading ? (
          <Container>
            <CachedIcon fontSize={'large'} /> Your cloud carbon footprint data is loading
          </Container>
        ) : (
          <ApexLineChart data={dataInTimeframe} />
        )}
      </Box>
    </>
  )
}

export default CloudCarbonContainer
