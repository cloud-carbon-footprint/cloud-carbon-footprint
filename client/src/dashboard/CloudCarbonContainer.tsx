import React from 'react'
import useRemoteService from './client/RemoteServiceHook'
import useFilters from './filters/FilterHook'
import { ApexLineChart } from './charts/ApexLineChart'
import { ApexDonutChart } from './charts/ApexDonutChart'
import { CarbonComparisonCard } from './CarbonComparisonCard'
import moment from 'moment'
import MonthFilter from './filters/MonthFilter'
import { Box, Card, CircularProgress, Grid } from '@material-ui/core'
import ServiceFilter from './filters/ServiceFilter'
import DateFilter from './filters/DateFilter'
import { makeStyles } from '@material-ui/core/styles'

const useStyles = makeStyles((theme) => ({
  boxContainer: {
    padding: theme.spacing(3, 10),
  },
  filterContainer: {
    display: 'flex',
    paddingBottom: theme.spacing(0.5),
  },
  filter: {
    marginRight: theme.spacing(0.5),
  },
}))

export default function CloudCarbonContainer() {
  const classes = useStyles()
  const startDate: moment.Moment = moment.utc().subtract(11, 'month')
  const endDate: moment.Moment = moment.utc()

  const { data, loading } = useRemoteService([], startDate, endDate)
  const { filteredData, filters, setFilters } = useFilters(data)

  return (
    <Box className={classes.boxContainer}>
      <Grid container>
        <Grid item xs={12}>
          <div className={classes.filterContainer}>
            {[ServiceFilter, DateFilter, MonthFilter].map((FilterComponent, i) => (
              <div key={i} className={classes.filter}>
                <FilterComponent filters={filters} setFilters={setFilters} />
              </div>
            ))}
          </div>
        </Grid>
        {loading ? (
          <CircularProgress />
        ) : (
          <Grid container spacing={3}>
            <Grid item xs={12}>
              <Card style={{ width: '100%', height: '100%' }}>
                <Box padding={3} paddingRight={4}>
                  <ApexLineChart data={filteredData} />
                  <Grid container justify="center">
                    <div>*estimated with average CPU not actual CPU</div>
                  </Grid>
                </Box>
              </Card>
            </Grid>
            <Grid item xs={6}>
              <CarbonComparisonCard data={filteredData} />
            </Grid>
            <Grid item xs={6}>
              <Card style={{ width: '100%', height: '100%' }}>
                <Box padding={3}>
                  <ApexDonutChart data={filteredData} />
                </Box>
              </Card>
            </Grid>
          </Grid>
        )}
      </Grid>
    </Box>
  )
}
