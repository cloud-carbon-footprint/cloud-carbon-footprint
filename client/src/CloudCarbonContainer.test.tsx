import React from 'react'
import { render } from "@testing-library/react";
import CloudCarbonContainer from './CloudCarbonContainer'
import { ApexLineChart } from "./ApexLineChart"
import useRemoteService from './hooks/RemoteServiceHook'
import generateEstimations from './data/generateEstimations'
import { ServiceResult, EstimationResult } from './types';
import moment from 'moment';

jest.mock("./ApexLineChart", () => ({
    ApexLineChart: jest.fn(() => null)
}))
jest.mock('./hooks/RemoteServiceHook')


const mockedUseRemoteService = useRemoteService as jest.MockedFunction<typeof useRemoteService>
const mockedApexLineChart = ApexLineChart as jest.Mocked<typeof ApexLineChart>

describe('CloudCarbonContainer', () => {
    let data: EstimationResult[];

    beforeEach(() => {
        data = generateEstimations(moment.utc(), 14)
        const mockReturnValue: ServiceResult = { loading: false, error: false, data: data }
        mockedUseRemoteService.mockReturnValue(mockReturnValue)
    })

    test("initial timeframe should be 12 months", () => {
        render(<CloudCarbonContainer />)

        expect(mockedApexLineChart).toBeCalledWith(
            {
                data: data.slice(0, 12)
            },
            expect.anything())
    });
});

