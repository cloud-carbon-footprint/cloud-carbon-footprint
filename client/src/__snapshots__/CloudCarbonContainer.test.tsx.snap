// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudCarbonContainer match against snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-1"
  >
    <div
      class="MuiButtonGroup-root"
      role="group"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          1M
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          3M
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          6M
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-containedPrimary MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          12M
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-2"
  />
</div>
`;
